### PR TITLE
Add Ledger API scala client for user management

### LINKED ISSUE #12009: scala bindings: create UserManagementServiceClient

### PR DESCRIPTION
**Pull request by meiersi-da**
Fixes #12009 

FYI: @cocreature @stefanobaghino-da @adriaanm-da 

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description
- [x] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala [hunk 1]
```diff
@@ -333,4 +337,18 @@ private[data] final class IdStringImpl extends IdString {
...
+   *  1. The characters `-._` never follow each other.
+   *  2. The characters `-._` neither occur at the start nor at the end of the user name.
+   *  Thus 'john.doe1' is a valid user-name, while 'john..doe1', 'john.-doe1', and '-john.doe' are not.
+   */
+  override type UserId = String
+  override val UserId: StringModule[UserId] =
+    new MatchingStringModule(
+      "User ID",
+      """[\p{Lower}\d]([\p{Lower}\d]|[-._][\p{Lower}\d]){0,62}"""
+      // FIXME: allow | as that one is used for example by Auth0 to prefix the identity provider (https://auth0.com/docs/users/user-profiles/sample-user-profiles). Doing this will though require introducing a separate string for [[ApplicationId]] to avoid unconditionally expanding the notion of a LedgerString.
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:351]: that seems like something we can do after merging this PR or potentially even after 2.0. Relaxing the format seems not-breaking.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 2]
```diff
@@ -47,6 +47,18 @@ object Ref {
...
+   *
+   * The concept of participant node users has been introduced after the concept
+   * of client applications, which is why the code contains both of them.
+   * The [[UserId]] identifiers for participant node users are a strict subset of the
+   * [[ApplicationId]] identifiers for Ledger API client applications.
+   * The Ledger API and its backing code use [[UserId]] where possible
+   * and [[ApplicationId]] otherwise.
+   * */
+  type UserId = IdString.UserId
+  val UserId: IdString.UserId.type = IdString.UserId
```
Comment (stefanobaghino-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:60]: If the concept of user never makes it into the evaluation of commands, I don't think that `daml-lf` is a good place for this.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala [hunk 3]
```diff
@@ -333,4 +337,18 @@ private[data] final class IdStringImpl extends IdString {
...
+   *  hyphens, underscores, and dots, and satisfy the following  rules:
+   *  1. The characters `-._` never follow each other.
+   *  2. The characters `-._` neither occur at the start nor at the end of the user name.
+   *  Thus 'john.doe1' is a valid user-name, while 'john..doe1', 'john.-doe1', and '-john.doe' are not.
+   */
+  override type UserId = String
+  override val UserId: StringModule[UserId] =
+    new MatchingStringModule(
+      "User ID",
+      """[\p{Lower}\d]([\p{Lower}\d]|[-._][\p{Lower}\d]){0,62}"""
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:350]: why 62?

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import com.daml.ledger.client.LedgerClient
+import com.daml.ledger.client.services.admin.UserManagementClient.{fromApiRight, fromProtoUser, toApiRight}
+import com.daml.lf.data.Ref
+import com.daml.lf.data.Ref.Party
+
+import scala.concurrent.{ExecutionContext, Future}
+
+final class UserManagementClient
+  (service: UserManagementServiceStub)
+  (implicit ec: ExecutionContext)
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala:17]: To save time, run `./fmt.sh --test --diff && ./ci/check-changelog.sh` on your machine before pushing. Or you can add it to a pre-push hook.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 5]
```diff
@@ -47,6 +47,18 @@ object Ref {
   type ApplicationId = LedgerString
   val ApplicationId: LedgerString.type = LedgerString
 
+  /** Identifiers for participant node users, which act as clients to the Ledger API.
+   *
+   * The concept of participant node users has been introduced after the concept
+   * of client applications, which is why the code contains both of them.
+   * The [[UserId]] identifiers for participant node users are a strict subset of the
```
Comment (cocreature) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:54]: I don’t understand the relation you are sketching out here. I thought your suggestion was that an application id identifies a full Daml App and all clients (which can map to different users) use the same one?

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  private def fromProtoUser(user: proto.User): User =
+    User(
+      UserId(Ref.UserId.assertFromString(user.id)),
+      Option.unless(user.primaryParty.isEmpty)(Party.assertFromString(user.primaryParty))
+    )
+
+  private def toProtoUser(user: User): proto.User =
+    proto.User(user.id.toString, user.primaryParty.fold("")(_.toString))
+
+  // FIXME: move to the right place in the codebase; and cleanup imports; and dedup with the code in 'ApiUserManagementService'
```
Comment (cocreature) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala:91]: I’m fine with addressing that after merging. At least within the language team we’ve found it useful to include a link to an issue for that type of stuff so you can easily grep for it later.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala [hunk 7]
```diff
@@ -0,0 +1,111 @@
+package com.daml.ledger.client.services.admin
+
+import com.daml.ledger.api.domain
+import com.daml.ledger.api.domain.{UserId, User, UserRight}
+import com.daml.ledger.api.v1.admin.user_management_service.UserManagementServiceGrpc.UserManagementServiceStub
+import com.daml.ledger.api.v1.admin.user_management_service.{GetUserRequest, Right}
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala:6]: Probably renaming `Right`  to `ApiRight` in the import would avoid confusion with the `Right` constructor of `Either`.

**Review by stefanobaghino-da — COMMENTED** [review]
Looks good so far, just a couple of comments of things to improve and look into as we move this into a mergeable state.

**Review by cocreature — APPROVED** [review]
nice, thank you!

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:60]: Same thing applies to `ApplicationId` which has been in here for ages. I’m fine with keeping all of them in here.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala (hunk 7 above)
Comment (cocreature) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala:6]: Or `UserRight`

Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala:6]: There already is a `UserRight` in scope.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala (hunk 3 above)
Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala:350]: Limit to at most 63 characters, which is a shorter variant of how we limit ledger strings to 255 characters.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 5 above)
Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala:54]: The initial idea was that `ApplicationId` becomes equivalent to `UserId` in our codebase. That doesn't really work, as then `UserId` have too wide a character set and in particular allow for spaces, which is a security problem wrt human redability.

So, the proposal here is that on top of the existing custom JWT tokens, which can be used to authenticate requests for any valid `ApplicationId`, the Ledger API server also allows standard JWT tokens whose `sub` claims must be a valid `UserId` to authenticate requests whose `application_id` field matches that `sub` claim.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala (hunk 7 above)
Comment (meiersi-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala]: Fixed.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala (hunk 4 above)
Comment (meiersi-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala]: Thx. Will do.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala (hunk 1 above)
Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala]: It is. I've created https://github.com/digital-asset/daml/issues/12001 to track this; and will remove the FIXME.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala (hunk 6 above)
Comment (meiersi-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/UserManagementClient.scala]: I've fixed this now.

**Review by remyhaemmerle-da — COMMENTED** [review]
LF changes look good.

**Review by cocreature — APPROVED** [review]
Thanks!

**Pull request merged**